### PR TITLE
Process API

### PR DESCRIPTION
**Pull request by bergie**
Refs noflo/noflo#410

### CODE REVIEW

**Comment by bergie** [comment]
OK, all components have tests and are on Process API now. @trustmaster ready for review

File: components/Drop.coffee [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
+  c.inPorts.add 'in',
+    datatypes: 'all'
+    description: 'Packet to be dropped'
 
-exports.getComponent = -> new Drop
+  c.process (input, output) ->
+    data = input.get 'in'
+    return unless data.type is 'data'
+    data.drop()
+    data.type = 'data'
```
Comment (trustmaster) [inline components/Drop.coffee]: Why does it need to do this after dropped?

Comment (bergie) [inline components/Drop.coffee]: Becauce otherwise the `data` packet has no type and it crashes Process API.

Comment (bergie) [inline components/Drop.coffee]: https://github.com/noflo/noflo/issues/446

File: components/MakeFunction.coffee [hunk 2]
```diff
@@ -1,61 +1,52 @@
...
-      if typeof data is "function"
-        @f = data
-      else
+  c.icon = 'code'
+
+  c.inPorts.add 'in',
+    datatype: 'all'
+    description: 'Packet to be processed'
+  c.inPorts.add 'function',
+    datatype: 'string'
```
Comment (trustmaster) [inline components/MakeFunction.coffee]: Doesn't this conflict with L28? Should be `'all'`?

Comment (bergie) [inline components/MakeFunction.coffee]: It kind of does. But the primary input is expected to be a string to evaluate. It is just that NoFlo doesn't actually enforce types, so we could just as well accept a real function

File: components/MakeFunction.coffee [hunk 3]
```diff
@@ -1,61 +1,52 @@
...
-
-    # Evaluate the function when receiving data
-    @inPorts.in.on 'data', (data) =>
-      unless @f
-        @error 'No function defined'
+          func = Function 'x', func
+        catch e
+          output.sendDone e
+          return
+      unless input.has 'in'
```
Comment (trustmaster) [inline components/MakeFunction.coffee]: What if `in` arrives a bit later than `function`? Then `function` would be sent as is and `in` would be queued until the next `function` arrives. Should probably be synchronized explicitly in the graph, but some might expect implicit synchronization here.

Comment (bergie) [inline components/MakeFunction.coffee]: It won't, since `function` is a control port

Comment (trustmaster) [inline components/MakeFunction.coffee]: `control` ports are `triggering: true` by default like any other ports, they just hold the last value received.

File: components/ReadEnv.coffee [hunk 4]
```diff
@@ -17,15 +17,14 @@ exports.getComponent = ->
...
-    return callback new Error "No environment variable #{data} set" if value is undefined
-    out.send process.env[data]
-    do callback
+  c.forwardBrackets =
+    key: ['out', 'error']
 
-  c
+  c.process (input, output) ->
+    data = input.get 'key'
+    value = process.env[data]
```
Comment (trustmaster) [inline components/ReadEnv.coffee]: `data` is supposed to be an `object` here

File: components/MakeFunction.coffee (hunk 3 above)
Comment (bergie) [inline components/MakeFunction.coffee]: Yep, in which case we just send the function out and wait for next input packet

File: components/RepeatDelayed.coffee [hunk 5]
```diff
@@ -14,22 +14,21 @@ exports.getComponent = ->
...
-    out: 'out'
-    forwardGroups: true
-    async: true
-  , (payload, groups, out, callback) ->
-    timer = setTimeout =>
-      out.send payload
-      do callback
+  c.process (input, output) ->
+    return unless input.has 'in'
+    delay = input.getData 'delay'
```
Comment (trustmaster) [inline components/RepeatDelayed.coffee]: Is `InPort.default` playing nicely with Process API here? I didn't test it in Component spec.

File: components/ReadEnv.coffee (hunk 4 above)
Comment (bergie) [inline components/ReadEnv.coffee]: Fixed in https://github.com/noflo/noflo-core/pull/78/commits/1bf14669f396a6437c98931d9bd04b47e9c8e55b

Was a bad test.

File: components/RepeatDelayed.coffee (hunk 5 above)
Comment (bergie) [inline components/RepeatDelayed.coffee]: It should get sent by Network.

Comment (trustmaster) [inline components/RepeatDelayed.coffee]: That's great.

**Comment by trustmaster** [comment]
@bergie nothing else suspicious to me.

**Pull request merged**